File: BinaryNotesMQ/java/src/org/bn/mq/impl/MQConnection.java
/*
 * Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
 * 
 * Licensed under the LGPL, Version 2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.gnu.org/copyleft/lgpl.html
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 * With any your questions welcome to my e-mail 
 * or blog at http://abdulla-a.blogspot.com.
 */

package org.bn.mq.impl;

import java.net.URI;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.bn.mq.IMQConnection;
import org.bn.mq.IMQConnectionListener;
import org.bn.mq.IPTPSession;
import org.bn.mq.IRemoteSupplier;
import org.bn.mq.ISupplier;
import org.bn.mq.net.ITransport;
import org.bn.mq.net.ITransportConnectionListener;
import org.bn.mq.net.ITransportReader;
import org.bn.mq.protocol.LookupRequest;
import org.bn.mq.protocol.LookupResult;
import org.bn.mq.protocol.LookupResultCode;
import org.bn.mq.protocol.MessageBody;
import org.bn.mq.protocol.MessageEnvelope;

public class MQConnection implements IMQConnection, ITransportConnectionListener, ITransportReader {
    protected ITransport transport;
    protected final int callTimeout = 30;
    protected Map<String,ISupplier> suppliers = new HashMap<String,ISupplier>();
    protected List < IMQConnectionListener> listeners = new LinkedList < IMQConnectionListener > ();
    
    public MQConnection(ITransport transport) {
        this.transport = transport;
        transport.addConnectionListener(this);
        transport.addReader(this);
    }
    
    public void start() throws Exception {
        this.transport.start();
    }

    public IRemoteSupplier lookup(String supplierName) throws Exception {
        LookupRequest request = new LookupRequest();
        request.setSupplierName(supplierName);
        MessageEnvelope message = new MessageEnvelope();
        MessageBody body = new MessageBody();
        body.selectLookupRequest(request);
        message.setBody(body);
        message.setId(this.toString());
        MessageEnvelope result = transport.call(message,callTimeout);
        if (result.getBody().getLookupResult().getCode().getValue() == LookupResultCode.EnumType.success ) {
            return new RemoteSupplier(supplierName,transport);
        }
        else
            throw new Exception("Error when accessing to supplier '"+supplierName+"': "+ result.getBody().getLookupResult().getCode().getValue().toString());
    }
  
    public ISupplier createSupplier(String supplierName) {
        Supplier supplier = new Supplier(supplierName,transport);
        synchronized(suppliers) {            
            suppliers.put(supplier.getId(),supplier);
        }        
        return supplier;
    }
    
    public void removeSupplier(ISupplier supplier) {
        synchronized(suppliers) {
            suppliers.remove(supplier.getId());
        }
    }    

    public URI getAddr() {
        return transport.getAddr();
    }

    public void close() {
        transport.delConnectionListener(this);
        transport.delReader(this);
        transport.close();
    }

    public boolean onReceive(MessageEnvelope message, ITransport replyTransport) {
        if(message.getBody().isLookupRequestSelected()) {
            MessageEnvelope result = new MessageEnvelope();
            result.setId(message.getId());
            MessageBody body = new MessageBody();
            result.setBody(body);
            LookupResult lResult = new LookupResult();
            body.selectLookupResult(lResult);
            try {
                synchronized(suppliers) {
                    ISupplier supplier = suppliers.get(message.getBody().getLookupRequest().getSupplierName());
                    LookupResultCode resCode = new LookupResultCode();                    
                    if(supplier!=null) {
                        resCode.setValue(LookupResultCode.EnumType.success);
                    }
                    else
                        resCode.setValue(LookupResultCode.EnumType.notFound);
                    lResult.setCode(resCode);
                }
                replyTransport.sendAsync(result);
            }
            catch (Exception e) {
                e.printStackTrace();
            }
            return true;
        }
        else
            return false;
    }

    public void onConnected(ITransport transport) {
        synchronized(listeners) {
            for(IMQConnectionListener listener: listeners) {
                listener.onConnected(this,transport);
            }
        }
    }

    public void onDisconnected(ITransport transport) {
        synchronized(listeners) {
            for(IMQConnectionListener listener: listeners) {
                listener.onDisconnected(this, transport);
            }
        }    
    }

    public void addListener(IMQConnectionListener listener) {
        synchronized(listeners) {
            listeners.add(listener);
        }
    }

    public void delListener(IMQConnectionListener listener) {
        synchronized(listeners) {
            listeners.remove(listener);
        }    
    }

    public <T> IPTPSession<T> createPTPSession(String pointName, String sessionName, Class<T> messageClass) {
        return new PTPSession<T>(pointName, sessionName, this.transport, messageClass);
    }
}

File: BinaryNotesMQ/.net/BinaryNotesMQ/src/org/bn/mq/impl/MQConnection.cs
/*
 Copyright 2006-2011 Abdulla Abdurakhmanov (abdulla@latestbit.com)
 Original sources are available at www.latestbit.com

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
using System;
using System.Collections.Generic;
using org.bn.mq.protocol;
using org.bn.mq.net;
using org.bn.mq;

namespace org.bn.mq.impl
{
	
	public class MQConnection : IMQConnection, ITransportConnectionListener, ITransportReader
	{
		virtual public Uri Addr
		{
			get
			{
				return transport.getAddr();
			}
			
		}
		protected internal ITransport transport;
		protected internal const int callTimeout = 30;
        protected internal IDictionary<String, ISupplier> suppliers = new Dictionary<String, ISupplier>();

        protected internal IList<IMQConnectionListener> listeners = new List<IMQConnectionListener>();
		
		public MQConnection(ITransport transport)
		{
			this.transport = transport;
			transport.addConnectionListener(this);
			transport.addReader(this);
		}
		
		public virtual IRemoteSupplier lookup(string supplierName)
		{

			MessageEnvelope message = new MessageEnvelope();
			MessageBody body = new MessageBody();

            LookupRequest request = new LookupRequest();
            request.SupplierName = supplierName;

			body.selectLookupRequest(request);
			message.Body = body;
			message.Id = this.ToString();
			MessageEnvelope result = transport.call(message, callTimeout);
			if (result.Body.LookupResult.Code.Value == LookupResultCode.EnumType.success)
			{
				return new RemoteSupplier(supplierName, transport);
			}
			else
				throw new System.Exception("Error when accessing to supplier '" + supplierName + "': " + result.Body.LookupResult.Code.Value.ToString());
		}
		
		public virtual ISupplier createSupplier(System.String supplierName)
		{
			Supplier supplier = new Supplier(supplierName, transport);
			lock (suppliers)
			{
				suppliers[supplier.Id] = supplier;
			}
			return supplier;
		}
		
		public virtual void  removeSupplier(ISupplier supplier)
		{
			lock (suppliers)
			{
				suppliers.Remove(supplier.Id);
			}
		}

        public void start()
        {
            transport.start();
        }
		
		public virtual void  close()
		{
			transport.delConnectionListener(this);
			transport.delReader(this);
			transport.close();
		}
		
		public virtual bool onReceive(MessageEnvelope message, ITransport replyTransport)
		{
			if (message.Body.isLookupRequestSelected())
			{
				MessageEnvelope result = new MessageEnvelope();
				result.Id = message.Id;
				MessageBody body = new MessageBody();
				result.Body = body;
				LookupResult lResult = new LookupResult();
				body.selectLookupResult(lResult);
				try
				{
					lock (suppliers)
					{
						ISupplier supplier = suppliers[message.Body.LookupRequest.SupplierName];
						LookupResultCode resCode = new LookupResultCode();
						if (supplier != null)
						{
							resCode.Value = LookupResultCode.EnumType.success;
						}
						else
							resCode.Value = LookupResultCode.EnumType.notFound;
						lResult.Code = resCode;
					}
					replyTransport.sendAsync(result);
				}
				catch (Exception e)
				{
                    Console.WriteLine(e);
                }
				return true;
			}
			else
				return false;
		}
		
		public virtual void  onConnected(ITransport transport)
		{
			lock (listeners)
			{
				foreach(IMQConnectionListener listener in listeners)
				{
					listener.onConnected(this, transport);
				}
			}
		}
		
		public virtual void  onDisconnected(ITransport transport)
		{
			lock (listeners)
			{
				foreach(IMQConnectionListener listener in listeners)
				{
					listener.onDisconnected(this, transport);
				}
			}
		}
		
		public virtual void  addListener(IMQConnectionListener listener)
		{
			lock (listeners)
			{
				listeners.Add(listener);
			}
		}
		
		public virtual void  delListener(IMQConnectionListener listener)
		{
			lock (listeners)
			{
				listeners.Remove(listener);
			}
		}

        public IPTPSession<T> createPTPSession<T>(String pointName, String sessionName)
        {
            return new PTPSession<T>(pointName,sessionName, transport);
        }
	}
}